### PR TITLE
Added node execution data to show inputs and outputs

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

## Node executions now show the input and outputs

```
- node_exec:
    closure:
      createdAt: "2021-07-08T14:25:58.559526900Z"
      phase: SUCCEEDED
      updatedAt: "2021-07-08T14:25:58.559526900Z"
    id:
      executionId:
        domain: development
        name: f0ecbec3960ce41188a2
        project: flytesnacks
      nodeId: start-node
    inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/start-node/data/inputs.pb
    metadata:
      specNodeId: start-node
- inputs:
    a: 3
  node_exec:
    closure:
      createdAt: "2021-07-08T14:25:58.603491Z"
      duration: 9.905055800s
      outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/node-t1-parent/data/0/outputs.pb
      phase: SUCCEEDED
      startedAt: "2021-07-08T14:25:58.667333600Z"
      updatedAt: "2021-07-08T14:26:08.572389800Z"
    id:
      executionId:
        domain: development
        name: f0ecbec3960ce41188a2
        project: flytesnacks
      nodeId: node-t1-parent
    inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/node-t1-parent/data/inputs.pb
    metadata:
      specNodeId: node-t1-parent
  outputs:
    c: world
    t1_int_output: 5
  task_execs:
  - closure:
      createdAt: "2021-07-08T14:25:58.657936Z"
      logs:
      - messageFormat: JSON
        name: Kubernetes Logs (User)
        uri: http://localhost:30082/#/log/flytesnacks-development/f0ecbec3960ce41188a2-node-t1-parent-0/pod?namespace=flytesnacks-development
      outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/node-t1-parent/data/0/outputs.pb
      phase: SUCCEEDED
      updatedAt: "2021-07-08T14:26:08.561845Z"
    id:
      nodeExecutionId:
        executionId:
          domain: development
          name: f0ecbec3960ce41188a2
          project: flytesnacks
        nodeId: node-t1-parent
      taskId:
        domain: development
        name: core.control_flow.subworkflows.t1
        project: flytesnacks
        resourceType: TASK
        version: v3
    inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/node-t1-parent/data/inputs.pb
- child_nodes:
  - inputs:
      o0: world
      o1: world
    node_exec:
      closure:
        createdAt: "2021-07-08T14:26:33.827371700Z"
        phase: SUCCEEDED
        updatedAt: "2021-07-08T14:26:33.906625900Z"
      id:
        executionId:
          domain: development
          name: f0ecbec3960ce41188a2
          project: flytesnacks
        nodeId: n1-0-end-node
      inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/end-node/inputs.pb
      metadata:
        retryGroup: "0"
        specNodeId: end-node
  - inputs:
      a: 5
    node_exec:
      closure:
        createdAt: "2021-07-08T14:26:08.903165200Z"
        duration: 4.652763200s
        outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n0/0/outputs.pb
        phase: SUCCEEDED
        startedAt: "2021-07-08T14:26:08.985713900Z"
        updatedAt: "2021-07-08T14:26:13.638477200Z"
      id:
        executionId:
          domain: development
          name: f0ecbec3960ce41188a2
          project: flytesnacks
        nodeId: n1-0-n0
      inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n0/inputs.pb
      metadata:
        retryGroup: "0"
        specNodeId: n0
    outputs:
      c: world
      t1_int_output: 7
    task_execs:
    - closure:
        createdAt: "2021-07-08T14:26:08.979419600Z"
        logs:
        - messageFormat: JSON
          name: Kubernetes Logs (User)
          uri: http://localhost:30082/#/log/flytesnacks-development/f0ecbec3960ce41188a2-n1-0-n0-0/pod?namespace=flytesnacks-development
        outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n0/0/outputs.pb
        phase: SUCCEEDED
        updatedAt: "2021-07-08T14:26:13.628872300Z"
      id:
        nodeExecutionId:
          executionId:
            domain: development
            name: f0ecbec3960ce41188a2
            project: flytesnacks
          nodeId: n1-0-n0
        taskId:
          domain: development
          name: core.control_flow.subworkflows.t1
          project: flytesnacks
          resourceType: TASK
          version: v3
      inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n0/inputs.pb
  - inputs:
      a: 7
    node_exec:
      closure:
        createdAt: "2021-07-08T14:26:13.842365600Z"
        duration: 19.748246700s
        outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n1/0/outputs.pb
        phase: SUCCEEDED
        startedAt: "2021-07-08T14:26:13.938074500Z"
        updatedAt: "2021-07-08T14:26:33.686320700Z"
      id:
        executionId:
          domain: development
          name: f0ecbec3960ce41188a2
          project: flytesnacks
        nodeId: n1-0-n1
      inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n1/inputs.pb
      metadata:
        retryGroup: "0"
        specNodeId: n1
    outputs:
      c: world
      t1_int_output: 9
    task_execs:
    - closure:
        createdAt: "2021-07-08T14:26:13.929258900Z"
        logs:
        - messageFormat: JSON
          name: Kubernetes Logs (User)
          uri: http://localhost:30082/#/log/flytesnacks-development/f0ecbec3960ce41188a2-n1-0-n1-0/pod?namespace=flytesnacks-development
        outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n1/0/outputs.pb
        phase: SUCCEEDED
        updatedAt: "2021-07-08T14:26:33.676085900Z"
      id:
        nodeExecutionId:
          executionId:
            domain: development
            name: f0ecbec3960ce41188a2
            project: flytesnacks
          nodeId: n1-0-n1
        taskId:
          domain: development
          name: core.control_flow.subworkflows.t1
          project: flytesnacks
          resourceType: TASK
          version: v3
      inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/n1/inputs.pb
  - node_exec:
      closure:
        createdAt: "2021-07-08T14:26:08.781564Z"
        phase: SUCCEEDED
        updatedAt: "2021-07-08T14:26:08.781564Z"
      id:
        executionId:
          domain: development
          name: f0ecbec3960ce41188a2
          project: flytesnacks
        nodeId: n1-0-start-node
      inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/start-node/inputs.pb
      metadata:
        retryGroup: "0"
        specNodeId: start-node
  node_exec:
    closure:
      createdAt: "2021-07-08T14:26:08.667611700Z"
      duration: 25.256306600s
      outputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/0/outputs.pb
      phase: SUCCEEDED
      startedAt: "2021-07-08T14:26:08.723882900Z"
      updatedAt: "2021-07-08T14:26:33.980189600Z"
    id:
      executionId:
        domain: development
        name: f0ecbec3960ce41188a2
        project: flytesnacks
      nodeId: n1
    inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/n1/data/inputs.pb
    metadata:
      isParentNode: true
      specNodeId: n1
- inputs:
    o0: 5
    o1: world
    o2: world
  node_exec:
    closure:
      createdAt: "2021-07-08T14:26:34.061023600Z"
      phase: SUCCEEDED
      updatedAt: "2021-07-08T14:26:34.133729200Z"
    id:
      executionId:
        domain: development
        name: f0ecbec3960ce41188a2
        project: flytesnacks
      nodeId: end-node
    inputUri: s3://my-s3-bucket/metadata/propeller/flytesnacks-development-f0ecbec3960ce41188a2/end-node/data/inputs.pb
    metadata:
      specNodeId: end-node
```

The default tree view .
```
flytectl get execution -p flytesnacks -d development --details f0ecbec3960ce41188a2 
TABLE format is not supported on detailed view and defaults to tree view. Choose either json/yaml

└── start-node - SUCCEEDED - 2021-07-08 14:25:58.5595269 +0000 UTC - 2021-07-08 14:25:58.5595269 +0000 UTC
└── node-t1-parent - SUCCEEDED - 2021-07-08 14:25:58.603491 +0000 UTC - 2021-07-08 14:26:08.5723898 +0000 UTC
│   ├── Attempt :0
│       └── Task - SUCCEEDED - 2021-07-08 14:25:58.657936 +0000 UTC - 2021-07-08 14:26:08.561845 +0000 UTC
│       └── Task Type - 
│       └── Reason - 
│       └── Logs :
│           └── Name :Kubernetes Logs (User)
│           └── URI :http://localhost:30082/#/log/flytesnacks-development/f0ecbec3960ce41188a2-node-t1-parent-0/pod?namespace=flytesnacks-development
└── n1 - SUCCEEDED - 2021-07-08 14:26:08.6676117 +0000 UTC - 2021-07-08 14:26:33.9801896 +0000 UTC
│   ├── n1-0-start-node - SUCCEEDED - 2021-07-08 14:26:08.781564 +0000 UTC - 2021-07-08 14:26:08.781564 +0000 UTC
│   ├── n1-0-n0 - SUCCEEDED - 2021-07-08 14:26:08.9031652 +0000 UTC - 2021-07-08 14:26:13.6384772 +0000 UTC
│   │   ├── Attempt :0
│   │       └── Task - SUCCEEDED - 2021-07-08 14:26:08.9794196 +0000 UTC - 2021-07-08 14:26:13.6288723 +0000 UTC
│   │       └── Task Type - 
│   │       └── Reason - 
│   │       └── Logs :
│   │           └── Name :Kubernetes Logs (User)
│   │           └── URI :http://localhost:30082/#/log/flytesnacks-development/f0ecbec3960ce41188a2-n1-0-n0-0/pod?namespace=flytesnacks-development
│   ├── n1-0-n1 - SUCCEEDED - 2021-07-08 14:26:13.8423656 +0000 UTC - 2021-07-08 14:26:33.6863207 +0000 UTC
│   │   ├── Attempt :0
│   │       └── Task - SUCCEEDED - 2021-07-08 14:26:13.9292589 +0000 UTC - 2021-07-08 14:26:33.6760859 +0000 UTC
│   │       └── Task Type - 
│   │       └── Reason - 
│   │       └── Logs :
│   │           └── Name :Kubernetes Logs (User)
│   │           └── URI :http://localhost:30082/#/log/flytesnacks-development/f0ecbec3960ce41188a2-n1-0-n1-0/pod?namespace=flytesnacks-development
│   ├── n1-0-end-node - SUCCEEDED - 2021-07-08 14:26:33.8273717 +0000 UTC - 2021-07-08 14:26:33.9066259 +0000 UTC
└── end-node - SUCCEEDED - 2021-07-08 14:26:34.0610236 +0000 UTC - 2021-07-08 14:26:34.1337292 +0000 UTC

```

# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
- [ ] Bug Fix
- [x] Feature
- [ ] Plugin

## Are all requirements met?

- [x] Code completed
- [x] Smoke tested
- [x] Unit tests added
- [x] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
_How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1124

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by kumare3** [comment]
@pmahindrakar-oss what happens if we specify
` flytectl get execution -p flytesnacks -d development --details fd44b4d331ae94d21a84 --nodeId coretypesystemcustomobjectsadd -o yaml`

**Comment by pmahindrakar-oss** [comment]
@kumare3 here the yaml o/p 
```
- created_at: "2021-07-08T14:31:47.2684791Z"
  ended_at: "2021-07-08T14:32:07.2439678Z"
  id:
    execution_id:
      domain: development
      name: f56557f1a3ecd4820959
      project: flytesnacks
    node_id: coretypesystemcustomobjectsadd
  inputs:
    x:
      x: 2
      "y": ydatafory
      z:
        "1": foo
        "2": bar
    "y":
      x: 3
      "y": ydataforx
      z:
        "3": buzz
        "4": lightyear
  outputs:
    o0:
      x: 5
      "y": ydataforyydataforx
      z:
        "1": foo
        "2": bar
        "3": buzz
        "4": lightyear
  phase: SUCCEEDED
  started_at: "2021-07-08T14:31:47.3175804Z"
  task_execs:
  - created_at: "2021-07-08T14:31:47.3060542Z"
    ended_at: "1970-01-01T00:00:00Z"
    id:
      node_execution_id:
        execution_id:
          domain: development
          name: f56557f1a3ecd4820959
          project: flytesnacks
        node_id: coretypesystemcustomobjectsadd
      task_id:
        domain: development
        name: core.type_system.custom_objects.add
        project: flytesnacks
        resource_type: 1
        version: v3
    logs:
    - message_format: 2
      name: Kubernetes Logs (User)
      uri: http://localhost:30082/#/log/flytesnacks-development/fmwt31ni/pod?namespace=flytesnacks-development
    phase: SUCCEEDED
    started_at: "1970-01-01T00:00:00Z"
    updated_at: "2021-07-08T14:32:07.2346061Z"
  updated_at: "2021-07-08T14:32:07.2439674Z"
```

File: cmd/get/execution.go [hunk 1]
```diff
@@ -76,6 +76,13 @@ Task execution view is also available in yaml/json format. Below example shows y
 
  bin/flytectl get execution -p flytesnacks -d development oeh94k9r2r --nodId n0 -o yaml
 
+Get the nodes input and outputs using --io flag
+
+::
+
+ bin/flytectl get execution -p flytesnacks -d development oeh94k9r2r --nodId n0 --io
```
Comment (kumare3) [inline cmd/get/execution.go:83]: thus this need to be nodeID

File: pkg/ext/execution_fetcher.go [hunk 2]
```diff
@@ -23,14 +22,32 @@ func (a *AdminFetcherExtClient) FetchExecution(ctx context.Context, name, projec
...
+func (a *AdminFetcherExtClient) FetchNodeExecutionDetails(ctx context.Context, name, project, domain, uniqueParentId string) (*admin.NodeExecutionList, error) {
 	ne, err := a.AdminServiceClient().ListNodeExecutions(ctx, &admin.NodeExecutionListRequest{
 		WorkflowExecutionId: &core.WorkflowExecutionIdentifier{
 			Project: project,
 			Domain:  domain,
 			Name:    name,
 		},
-		Limit: 100,
+		UniqueParentId: uniqueParentId,
+		Limit:          100,
```
Comment (kumare3) [inline pkg/ext/execution_fetcher.go:50]: what if number of nodes is more than 100?

File: pkg/printer/printer.go [hunk 3]
```diff
@@ -124,6 +124,39 @@ func (p Printer) JSONToTable(jsonRows []byte, columns []Column) error {
 	return nil
 }
 
+// TODO : refactor to reuse
+func (p Printer) PrintInterface(format OutputFormat, columns []Column, v interface{}) error {
```
Comment (kumare3) [inline pkg/printer/printer.go:128]: can we do this refactoring?

**Comment by kumare3** [comment]
some unit tests are failing! Once they are fixed +1

File: pkg/ext/execution_fetcher.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply pkg/ext/execution_fetcher.go:50]: We are going to support pagination across all flyte entities where we will address this 
https://github.com/flyteorg/flyte/issues/1146

File: cmd/get/execution.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply cmd/get/execution.go]: Done

File: pkg/printer/printer.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply pkg/printer/printer.go:127]: Done

File: cmd/get/node_execution.go [hunk 4]
```diff
@@ -52,151 +43,239 @@ const (
...
+	// Time at which the node execution was last updated.
+	UpdatedAt time.Time `json:"updated_at,omitempty"`
+}
 
-	// Fetching Node execution details
-	nExecDetails, nodeExecToTaskExec, err := getNodeExecDetailsWithTasks(ctx, project, domain, name, cmdCtx)
-	if err != nil {
-		return err
-	}
+type NodeExecutionWrapper struct {
```
Comment (kumare3) [inline cmd/get/node_execution.go:79]: Lets add comments for these structs, to explain - why do we need them, whats the goal

Comment (kumare3) [reply cmd/get/node_execution.go:79]:
```golang
type NodeExecutionClosure struct {
*NodeExecution
ChildNodes []*NodeExecutionWrapper `json:"child_nodes,omitempty"`
TaskExecutions []*TaskExecutionWrapper `json:"task_execs,omitempty"
}
```

Comment (pmahindrakar-oss) [reply cmd/get/node_execution.go:79]: Hi @kumare3 , i will take these as a followup to this PR. facing issues with the marshalling of these Closure classes and needs to dug deeper.
For now we can go with these wrapper classes and i will soon send followup PR with this fix.

Comment (pmahindrakar-oss) [reply cmd/get/node_execution.go]: Done

**Review by kumare3 — APPROVED** [review]

**Comment by kumare3** [comment]
Great job, thank you!
